lib/createReconciler.lua

```diff
@@ -0,0 +1,239 @@
+local Type = require(script.Parent.Type)
+local ElementKind = require(script.Parent.ElementKind)
+local Core = require(script.Parent.Core)
+
+local function noop()
+	return nil
+end
+
+local function iterateElements(childOrChildren)
+	local richType = Type.of(childOrChildren)
+
+	-- Single child, the simplest case!
+	if richType == Type.Element then
+		local called = false
+
+		return function()
+			if called then
+				return nil
+			else
+				called = true
+				return 1, childOrChildren
+			end
+		end
+	end
+
+	-- This is a Roact-speciifc object, and it's the wrong kind.
+	if richType ~= nil then
+		error("Invalid children")
+	end
+
+	local regularType = typeof(childOrChildren)
+
+	-- A dictionary of children, hopefully!
+	-- TODO: Is this too flaky? Should we introduce a Fragment type like React?
+	if regularType == "table" then
+		return pairs(childOrChildren)
+	end
+
+	if childOrChildren == nil or regularType == "boolean" then
+		return noop
+	end
+
+	error("Invalid children")
+end
+
+local function createReconciler(renderer)
+	local reconciler
+	local mountNode
+	local reconcileNode
+
+	local function reconcileChildren(node, newChildElements)
+		for key, newElement in iterateElements(newChildElements) do
+			local node = node.children[key]
+
+			if node ~= nil then
+				node.children[key] = reconcileNode(node, newElement)
```

Comment:
I think your naming is overlapping here?

Reply:
Good catch, I haven't run this code yet. I'll update the `.luacheckrc` to be stricter too, since it normally would catch this problem.